### ISSUE TITLE
[Bug]: Rate Limit Errors with models other than OAI

### ISSUE THREAD

**Issue by gurugecl** [BerriAI/litellm#7358]
### What happened?

Starting 12/20/24 I keep getting rate limit errors when using any model other than OAI including Sonnet and Gemini Flash even though I am well below the rate limit. Previously this was working fine. I am using LiteLLM via PaperQA 

https://github.com/Future-House/paper-qa

Below is how I am setting up the Settings object which then throws the rate limit error. 

```
from paperqa import Docs, Settings

settings = Settings(
        llm="claude-3-5-sonnet-20241022",
        summary_llm="claude-3-5-sonnet-20241022",
        llm_config={
            "model_list": [{
                "model_name": "claude-3-5-sonnet-20241022",
                "litellm_params": {
                    "model": "claude-3-5-sonnet-20241022",
                    "api_key": os.environ.get('ANTHROPIC_API_KEY'),
                }
            }]
        },
        summary_llm_config={
            "model_list": [{
                "model_name": "claude-3-5-sonnet-20241022",
                "litellm_params": {
                    "model": "claude-3-5-sonnet-20241022",
                    "api_key": os.environ.get('ANTHROPIC_API_KEY'),
                }
            }]
        }
    ) 

max_choices = len(list(docs.docnames))
settings.answer.answer_max_sources = max_choices
settings.answer.evidence_k = relevancy * max_choices

model_response = docs.query(model_input, settings=settings)
```

### Relevant log output

```shell
LiteLLM completion() model= claude-3-5-sonnet-20241022; provider = anthropic
HTTP Request: POST https://api.anthropic.com/v1/messages "HTTP/1.1 429 Too Many Requests"
03:12:39 - LiteLLM Router:INFO: router.py:849 - litellm.acompletion(model=claude-3-5-sonnet-20241022) Exception litellm.RateLimitError: AnthropicException - {"type":"error","error":{"type":"rate_limit_error","message":"This request would exceed your organization’s rate limit of 8,000 output tokens per minute. For details, refer to: https://docs.anthropic.com/en/api/rate-limits; see the response headers for current usage. Please reduce the prompt length or the maximum tokens requested, or try again later. You may also contact sales at https://www.anthropic.com/contact-sales to discuss your options for a rate limit increase."}}
litellm.acompletion(model=claude-3-5-sonnet-20241022) Exception litellm.RateLimitError: AnthropicException - {"type":"error","error":{"type":"rate_limit_error","message":"This request would exceed your organization’s rate limit of 8,000 output tokens per minute. For details, refer to: https://docs.anthropic.com/en/api/rate-limits; see the response headers for current usage. Please reduce the prompt length or the maximum tokens requested, or try again later. You may also contact sales at https://www.anthropic.com/contact-sales to discuss your options for a rate limit increase."}}
```

### Are you a ML Ops Team?

Yes

### What LiteLLM version are you on ?

1.45.0

**Comment by ghost** [comment]
hi @gurugecl 

>  "error": {
    "code": 429,
    "message": "Resource has been exhausted (e.g. check quota).",
    "status": "RESOURCE_EXHAUSTED"
  }

your error is coming from google saying your resource is exhausted. This doesn't even look like it's coming from the router, just from the backend llm api. Closing for now as i don't see a litellm-specific issue, cc: @jamesbraza let me know if i'm missing something here

**Comment by gurugecl** [comment]
Yes but the issue seems to be that the error is coming on the very first request to Gemini and no amount of exponential backoff or waiting seems to be helping which I believe would indicate something is happening behind the scenes either with Paperqa or Litellm to exhaust the resource so quickly. The following seems to indicate that Gemini Flash 2.0 is supported [7188](https://github.com/BerriAI/litellm/issues/7188) and I even tried setting rate limits in the llm config like below according to the PaperQA docs but that did not seem to help

```
"rate_limit": {
                "gemini/gemini-2.0-flash-exp": "10 per 1 minute"  # Gemini's actual rate limit
            }
```

However if this not a LiteLLM issue I can reach out to PaperQA for further assistance

**Comment by ghost** [comment]
@gurugecl you can validate this with a simple test with litellm 

```
from litellm import completion
import os

os.environ['GEMINI_API_KEY'] = ""
response = completion(
    model="gemini/gemini-2.0-flash-exp", 
    messages=[{"role": "user", "content": "write code for saying hi from LiteLLM"}]
)
```

cc: @jamesbraza - let me know if i'm missing something here

**Comment by gurugecl** [comment]
Yes that works fine as does using Gemini through the API directly. However the problem arises when Im using Gemini for PaperQA which in turn uses LiteLLM I believe as I describe in this issue I just opened with them, [786](https://github.com/Future-House/paper-qa/issues/786). I'm using their Docs object as I need to manually add papers, however most of their documentation is in regards to the ask function for querying papers not the Docs class so Im not sure if the rate limit can be set in the same way and setting the max_concurrent_requests as they suggested didn't resolve the issues yet either

**Comment by ghost** [comment]
@gurugecl does it work when trying to use the openai api? with them

**Comment by ghost** [comment]
if so - one way around this is to spin up the cli proxy

```bash
export GEMINI_API_KEY=<your-key>
litellm --model gemini/gemini-2.0-flash-exp

# RUNNING on http://0.0.0.0:4000
```

and then set the openai api base from paperqa to this ^ 

```
export OPENAI_API_BASE="http://0.0.0.0:4000"
```

More docs - https://docs.litellm.ai/docs/#quick-start-proxy---cli

**Comment by gurugecl** [comment]
Yes it works fine with the OAI api as I believe it is used by default. Unfortunately, I couldn't quite get the proxy to work due to additional errors however I was able to get the implementation working without it by modifying the setup of the Docs object. Not sure yet why original approach worked initially but thanks a lot for your assistance with this. Really appreciate it!

**Comment by ghost** [comment]
Nice. Curious what were the proxy errors?

**Comment by ghost** [comment]
bump on this? @gurugecl 

would be helpful so i know if there's a key issue we might've missed

**Comment by gurugecl** [comment]
So Im not sure how much of an issue this is on your end but after setting up everything as we discussed I received the following error when trying to add new papers to the PaperQA Docs object:

```
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

Failed to load ddu148.pdf: litellm.BadRequestError: OpenAIException - Error code: 400 - {'error': {'message': "litellm.UnsupportedParamsError: Setting {'encoding_format': 'base64'} is not supported by gemini. To drop it from the call, set `litellm.drop_params = True`.", 'type': 'None', 'param': None, 'code': '400'}} LiteLLM Retried: 3 times

Loaded 0 documents
```

minimal script:
```
from paperqa import Docs, Settings
import os
from pathlib import Path
import litellm

# Set LiteLLM verbose logging
litellm.set_verbose = True

# Set up environment
os.environ["OPENAI_API_BASE"] = "http://0.0.0.0:4000"

# Initialize Docs
docs = Docs()

# Load PDFs from a directory
pdf_dir = Path("test_pdfs") 
for pdf_file in pdf_dir.glob("*.pdf"):
    print(f"Loading {pdf_file.name}")
    try:
        docs.add(str(pdf_file))
    except Exception as e:
        print(f"Failed to load {pdf_file.name}: {str(e)}")
```

So I modified the litellm command to run like this:

`litellm --model gemini/gemini-2.0-flash-exp --drop_params`

However at that point it just hung when trying to add papers and that's when I was able to resolve the issues without using the proxy as I mentioned so I didn't attempt to debug the proxy issue beyond this point

**Comment by jamesbraza** [comment]
It looks like you've misconfigured `Settings` a bit based on the Pydantic errors, but the more interesting one to me is this one: `Failed to load ddu148.pdf: litellm.BadRequestError: OpenAIException - Error code: 400 - {'error': {'message': "litellm.UnsupportedParamsError: Setting {'encoding_format': 'base64'} is not supported by gemini.`

I am interested where the `base64` part comes from, it's not within `paper-qa`

**Comment by ghost** [comment]
i believe that value is set when calling openai embeddings (i believe on their sdk, although i can check that)

**Comment by ghost** [comment]
nope, their default is float 

<img width="824" alt="Screenshot 2025-01-04 at 5 10 34 PM" src="https://github.com/user-attachments/assets/e5b0d226-0cff-4683-91dc-020ddfb7a53f" />

--
how do you handle embedding calls? @jamesbraza

**Comment by gurugecl** [comment]
@jamesbraza yup, sorry, the settings errors are what I fixed to get it to work without the proxy and not related to the proxy issue. Yeah it's the other base64 error that comes up when using the proxy and adding new papers to the Docs object

**Comment by jamesbraza** [comment]
> -- how do you handle embedding calls? @jamesbraza

We call `litellm.aembedding`: https://github.com/Future-House/llm-client/blob/v0.0.7/llmclient/embeddings.py#L149-L154

---

And I guess it seems from `Setting {'encoding_format': 'base64'} is not supported by gemini` it seems `base64` is not supported by Gemini. Maybe `base64` shouldn't be the default for all model providers?

**Comment by ghost** [comment]
> Maybe base64 shouldn't be the default for all model providers?

where do you see this value being set? @jamesbraza 

in general, yes - i don't think it needs to be unless there's a specific need for it

**Comment by jamesbraza** [comment]
It looks `"base64"` gets set within OpenAI's client library during `httpx` request creation: https://github.com/openai/openai-python/blob/v1.58.1/src/openai/resources/embeddings.py#L217

I wonder if something in @gurugecl 's code or LiteLLM `embedding` is incorrectly causing an OpenAI control flow to be used for a Gemini model. It could explain why `{'encoding_format': 'base64'}` is being passed to Gemini

**Comment by ghost** [comment]
@jamesbraza It was when he tried to call Gemini via the cli proxy using the openai route in the paperqa code

**Comment by gurugecl** [comment]
> It looks `"base64"` gets set within OpenAI's client library during `httpx` request creation: https://github.com/openai/openai-python/blob/v1.58.1/src/openai/resources/embeddings.py#L217
> 
> I wonder if something in @gurugecl 's code or LiteLLM `embedding` is incorrectly causing an OpenAI control flow to be used for a Gemini model. It could explain why `{'encoding_format': 'base64'}` is being passed to Gemini

Yeah the code that throws the error just consists of the below minimum example 

https://github.com/BerriAI/litellm/issues/7358#issuecomment-2571406172

plus running the proxy commands I was given in a separate terminal 

https://github.com/BerriAI/litellm/issues/7358#issuecomment-2569959857
